pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,243 @@ package heal
 
 import (
 	"context"
-
 	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
 	"google.golang.org/grpc"
+	"sync"
+	"time"
 
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 )
 
-// RegisterClientFunc - required to inform heal server about new client connection and assign it to the connection ID
-type RegisterClientFunc func(context.Context, *networkservice.Connection, networkservice.MonitorConnectionClient)
+type successVerifyRequest struct {
+	ch         chan *networkservice.PathSegment
+	prefixConn *networkservice.Connection
+}
 
 type healClient struct {
-	ctx context.Context
-	cc  networkservice.MonitorConnectionClient
+	ctx      context.Context
+	cc       networkservice.MonitorConnectionClient
+	initOnce sync.Once
+	initErr  error
+
+	// connCache contains connections which IDs we got in Request() updated to actual values as soon as monitor tells us something's changed
+	connCache      map[string]*networkservice.Connection
+	connCacheMutex sync.Mutex
+
+	// expectedConns contains connections we will be waiting to happen
+	expectedConns      map[string]successVerifyRequest
+	expectedConnsMutex sync.Mutex
 }
 
 // NewClient - creates a new networkservice.NetworkServiceClient chain element that inform healServer about new client connection
 func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient) networkservice.NetworkServiceClient {
 	return &healClient{
-		ctx: ctx,
-		cc:  cc,
+		ctx:           ctx,
+		cc:            cc,
+		connCache:     make(map[string]*networkservice.Connection),
+		expectedConns: make(map[string]successVerifyRequest),
 	}
 }
 
 func (u *healClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	u.initOnce.Do(func() {
+		errCh := make(chan error, 1)
+		pushFunc := healRequestFunc(ctx)
+		go u.listenToConnectionChanges(pushFunc, errCh)
+		u.initErr = <-errCh
+	})
+	// if initialization failed, then we want for all subsequent calls to Request() on this object to also fail
+	if u.initErr != nil {
+		return nil, u.initErr
+	}
+
+	// TODO what if this is a second call to Request() with this conn id?
+	//  will we receive a new event with this channel?
+	//  If not, then we will block on waiting and return an error eventually
+	verifySuccessCh, err := u.addToExpectedConnections(request.GetConnection())
+	if err != nil {
+		return nil, err
+	}
+
 	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return conn, err
```

Comment:
We have the following pattern for such case:
```go
func Request() {
    conn, err := next.Request()
    if err != nil {
        return nil, err // <-- return `nil` instead of `conn`
    }
    err := doSomething()
    if err != nil {
        next.Close(conn) // <-- close `conn` if we failed to process it
        return nil, err // <-- return `nil` instead of `conn`
    }
    return conn, nil
}
```